### PR TITLE
Update grpcio to 0.10

### LINKED ISSUE #1577: Update grpcio
Update the version of [`grpcio`](https://docs.rs/grpcio) we use. Ideally we'd move away from maintaining [our own fork](https://github.com/mobilecoinofficial/grpc-rs).

May benefit from #1408

### PR DESCRIPTION
**Pull request by remoun**
### Motivation
Fixes #1577.

### In this PR
* Update grpcio to 0.10 everywhere
* DRY up some of the `update_from_server_metadata` calls.

This is similar to #1144, and supersedes it.

### CODE REVIEW

**Review by jcape — CHANGES_REQUESTED** [review]
Some nitpicks in `thick.rs`, otherwise GTG

File: connection/src/thick.rs [hunk 1]
```diff
@@ -169,40 +169,56 @@ impl<CP: CredentialsProvider> ThickClient<CP> {
...
+                    )
+                }
+
+                response
+            })
+            .map_err(|err| {
+                let err = ThickClientAttestationError::from(err);
+                self.reset_if_unauthenticated(&err);
+                err
+            })
```
[the remark this answers is not in the document]
Comment (remoun) [reply connection/src/thick.rs:208]: I tried this, but it didn't work -- `map` and `map_err` return `Result`, while `map_or_else` only returns the `Ok` type.

[the remark this answers is not in the document]
Comment (jcape) [reply connection/src/thick.rs:210]: Ah, OK

**Review by jcape — APPROVED** [review]

File: connection/src/thick.rs [hunk 2]
```diff
@@ -304,23 +318,8 @@ impl<CP: CredentialsProvider> BlockchainConnection for ThickClient<CP> {
...
-                log::warn!(
-                    this.logger,
-                    "Could not update cookies from gRPC metadata: {}",
-                    e
-                )
-            }
-
-            Ok(message)
+            this.blockchain_api_client
+                .get_blocks_async_opt(&request, call_option)
```
Comment (iamalwaysuncomfortable) [inline connection/src/thick.rs:322]: Is changing this to the async opt going to break how blocks are currently syncing for existing clients? Like should we test how the various clients sync (I'm unsure whether test coverage covers this)

File: fog/api/src/lib.rs [hunk 3]
```diff
@@ -68,15 +68,15 @@ impl EnclaveGrpcChannel for ledger_grpc::FogKeyImageApiClient {
...
+        <Self>::auth_async_opt(self, msg, call_option)?.receive_sync()
     }
     fn enclave_request(
         &mut self,
         msg: &attest::Message,
         call_option: CallOption,
-    ) -> Result<(Option<Metadata>, attest::Message, Option<Metadata>), grpcio::Error> {
-        <Self>::check_key_images_full(self, msg, call_option)
+    ) -> GrpcResult<(Metadata, attest::Message, Metadata)> {
+        <Self>::check_key_images_async_opt(self, msg, call_option)?.receive_sync()
```
Comment (iamalwaysuncomfortable) [inline fog/api/src/lib.rs:79]: Again here, this looks like a case of "should be fine, but we should acceptance test it before the release is made"

File: connection/src/thick.rs [hunk 4]
```diff
@@ -87,9 +90,9 @@ impl From<CipherError> for ThickClientAttestationError {
 
 impl From<UriConversionError> for ThickClientAttestationError {
     fn from(src: UriConversionError) -> Self {
-        match src.clone() {
+        match &src {
```
Comment (iamalwaysuncomfortable) [inline connection/src/thick.rs:93]: Good catch!

**Review by iamalwaysuncomfortable — APPROVED** [review]

**Comment by iamalwaysuncomfortable** [comment]
Looks good although the switch to the async_opts in the GRPC api might lead to unexpected behavior (like making a ton of calls maybe?) to the block streaming becoming ready. Worth possibly some more test coverage prior to release.

File: fog/api/src/lib.rs (hunk 3 above)
Comment (remoun) [reply fog/api/src/lib.rs:79]: The `_full` suffix was our proposal, implemented only in our fork. The functionality has been implemented in the published crate with the `_async_opt` suffix.

File: connection/src/thick.rs (hunk 2 above)
Comment (remoun) [reply connection/src/thick.rs:322]: The `_full` suffix was our proposal, implemented only in our fork. The functionality has been implemented in the published crate with the `_async_opt` suffix.

**Comment by remoun** [comment]
> Looks good/approved although the switch to the async_opts in the GRPC api might lead to unexpected behavior (like making a ton of calls maybe?) to the block streaming becoming ready. Worth possibly some more test coverage prior to release.

Fair point. The `_full` suffix was our proposal, implemented only in our fork. The functionality has been implemented in the published crate with the `_async_opt` suffix.

**Pull request merged**